zcash_client_sqlite/src/for_async/mod.rs

```diff
@@ -0,0 +1,488 @@
+pub mod init;
+pub mod wallet_actions;
+
+use std::collections::HashMap;
+use std::path::Path;
+use zcash_client_backend::data_api::{PrunedBlock, ReceivedTransaction, SentTransaction};
+use zcash_client_backend::wallet::{AccountId, SpendableNote};
+use zcash_extras::{WalletRead, WalletWrite};
+use zcash_primitives::block::BlockHash;
+use zcash_primitives::consensus::BlockHeight;
+use zcash_primitives::memo::Memo;
+use zcash_primitives::merkle_tree::{CommitmentTree, IncrementalWitness};
+use zcash_primitives::sapling::{Node, Nullifier, PaymentAddress};
+use zcash_primitives::transaction::components::Amount;
+use zcash_primitives::transaction::TxId;
+use zcash_primitives::zip32::ExtendedFullViewingKey;
+
+pub async fn async_blocking<F, R>(blocking_fn: F) -> R
+where
+    F: FnOnce() -> R + Send + 'static,
+    R: Send + 'static,
+{
+    tokio::task::spawn_blocking(blocking_fn)
+        .await
+        .expect("spawn_blocking to succeed")
+}
+
+use crate::error::SqliteClientError;
+use crate::{wallet, NoteId, WalletDb};
+use rusqlite::Connection;
+use std::sync::{Arc, Mutex};
+
+use zcash_primitives::consensus;
+
+/// A wrapper for the SQLite connection to the wallet database.
+#[derive(Clone)]
+pub struct WalletDbAsync<P> {
+    inner: Arc<Mutex<WalletDb<P>>>,
+}
```

Comment:
Given all the above comments, I am actually wondering if `WalletDbAsync` and `DataConnStmtCacheAsync` can be implemented in mm2 instead of `librustzcash`. But don't worry about that for now @borngraced as we should continue reviewing and improving the code here. In the end, after making the changes in this PR optimal, we can look for the best place for it.